core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -1373,6 +1384,9 @@ private[kafka] class Processor(
     try {
       beginShutdown()
       thread.join()
+      if (!started) {
```

Comment:
When `socketServer` executes `close`, if `started` of `socketServer` is true, then the logic of `closeAll` can be executed in finally. If `started` of `socketServer` is false, then it is also executed the logic of `closeAll` at line 683 , so for the `close` of the processor, is there still a redundant call to `closeAll`?

Reply:
I don't think I understand what you mean. There's two parallel fixes here Acceptor and Processor, they both have the same flaw, and the flaw needs to be fixed in both places to prevent leaks.

I'm using synonymous variable and method names names, which is confusing because this is all in one file:
* Acceptor close
* Acceptor closeAll
* Acceptor started
* Processor close
* Processor closeAll
* Processor started

The SocketServer#shutdown calls Acceptor#beginShutdown, Acceptor#close, Processor#beginShutdown, and Acceptor#close. Internally, the Acceptor and Processor keep track of the started variable, and call closeAll exactly once: In finally if the thread was started, and in close if the thread was not.

Reply:
> I don't think I understand what you mean. There's two parallel fixes here Acceptor and Processor, they both have the same flaw, and the flaw needs to be fixed in both places to prevent leaks.
> 
> I'm using synonymous variable and method names names, which is confusing because this is all in one file:
> 
> * Acceptor close
> * Acceptor closeAll
> * Acceptor started
> * Processor close
> * Processor closeAll
> * Processor started
> 
> The SocketServer#shutdown calls Acceptor#beginShutdown, Acceptor#close, Processor#beginShutdown, and Processor#close. Internally, the Acceptor and Processor each keep track of a separate started variable, and call each closeAll exactly once: In finally if the thread was started, and in close if the thread was not.

What I mean is: when the thread of `SocketServer` does not get up, `closeAll` will be executed through `started` as false, then there is no need to execute `closeAll` once more in `Processor`, even if `Processor` `started` in ` is also false? @gharris1727

Reply:
> What I mean is: when the thread of SocketServer does not get up, closeAll will be executed through started as false

SocketServer doesn't have a closeAll or started field, it's in the Acceptor.

> then there is no need to execute closeAll once more in Processor, even if Processor started in ` is also false?

Processor has a selector that needs to be closed, even if there are no sockets attached.

I think I'm still not grasping what you're saying. Can you be more specific with which function/line numbers you think are run twice?

Reply:
> what I mean is: serverChannel and nioSelector are closed in closeAll. There only be one of these two objects in an Acceptor

I think an Acceptor will always have an nioSelector, and has a serverChannel sometimes before Acceptor#start is called, and always after Acceptor#start is called. That seems irrelevant to the Processor discussion though.

> So, is there no need to do extra closeAll at the Processor level?

When Processor is instantiated, Processor#selector is initialized. This is closed in Processor#stopAll, so we need to ensure that Processor#stopAll is called exactly once. It's not a ServerSocketChannel, but it should still be closed like the sockets are.
There are two calls to Processor#closeAll, (A) the Processor#run finally clause, and (B) the new Processor#close call that I added.

1. If Processor#start is called, the Processor#thread will start, and Processor#run will eventually exit, call site (A) will be executed. This is the existing happy-path. When Processor#close is called, call site (B) will be skipped because the thread was started.

2. If Processor#start is not called, the Processor#thread will not start, and call site (1) will never activate. Instead call site (2) will activate, because Processor#started will be false.

If you're asking why we don't remove call site (A) and make call site (B) unconditional: I didn't want to change the existing happy-path behavior.

If this isn't satisfactory I'll take some time to re-read your messages and put together a convincing test that demonstrates the need for the Processor change.